server/service/orbit.go

```diff
@@ -0,0 +1,138 @@
+)
+
+type orbitError struct {
+	message     string
+	nodeInvalid bool
+}
+
+func (e orbitError) Error() string {
+	return e.message
+}
+
+func (e orbitError) NodeInvalid() bool {
+	return e.nodeInvalid
+}
+
+func (r enrollOrbitResponse) error() error { return r.Err }
+
+func enrollOrbitEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+	req := request.(*enrollOrbitRequest)
+	nodeKey, err := svc.EnrollOrbit(ctx, req.HardwareUUID, req.EnrollSecret)
+	if err != nil {
+		return enrollOrbitResponse{Err: err}, nil
+	}
+	return enrollOrbitResponse{OrbitNodeKey: nodeKey}, nil
+}
+
+func (svc *Service) AuthenticateOrbitHost(ctx context.Context, orbitNodeKey string) (*fleet.Host, bool, error) {
+	svc.authz.SkipAuthorization(ctx)
+
+	if orbitNodeKey == "" {
+		return nil, false, ctxerr.Wrap(ctx, fleet.NewAuthRequiredError("authentication error: missing orbit node key"))
+	}
+
+	host, err := svc.ds.LoadHostByOrbitNodeKey(ctx, orbitNodeKey)
+	switch {
+	case err == nil:
+		// OK
+	case fleet.IsNotFound(err):
+		return nil, false, ctxerr.Wrap(ctx, fleet.NewAuthRequiredError("authentication error: invalid orbit node key"))
+	default:
+		return nil, false, ctxerr.Wrap(ctx, err, "authentication error orbit")
+	}
+
+	return host, svc.debugEnabledForHost(ctx, host.ID), nil
+}
+
+// EnrollOrbit returns an orbit nodeKey on successful enroll
+func (svc *Service) EnrollOrbit(ctx context.Context, hardwareUUID string, enrollSecret string) (string, error) {
+	// this is not a user-authenticated endpoint
+	svc.authz.SkipAuthorization(ctx)
+	logging.WithExtras(ctx, "hardware_uuid", hardwareUUID)
+
+	_, err := svc.ds.VerifyEnrollSecret(ctx, enrollSecret)
+	if err != nil {
+		return "", orbitError{
+			message:     "orbit enroll failed: " + err.Error(),
+			nodeInvalid: true,
+		}
+	}
+
+	orbitNodeKey, err := server.GenerateRandomText(svc.config.Osquery.NodeKeySize)
+	if err != nil {
+		return "", orbitError{
+			message:     "failed to generate orbit node key: " + err.Error(),
+			nodeInvalid: true,
+		}
+	}
+
+	_, err = svc.ds.EnrollOrbit(ctx, hardwareUUID, orbitNodeKey)
+	if err != nil {
+		return "", orbitError{
+			message:     "failed to enroll " + err.Error(),
+			nodeInvalid: true,
+		}
+	}
+
+	return orbitNodeKey, nil
+}
+
+func getOrbitFlagsEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+	req := request.(*orbitGetConfigRequest)
+	opts, err := svc.GetOrbitFlags(ctx, req.OrbitNodeKey)
+	if err != nil {
+		return enrollOrbitResponse{Err: err}, nil
+	}
+	return opts, nil
+}
+
+func (svc *Service) GetOrbitFlags(ctx context.Context, orbitNodeKey string) (json.RawMessage, error) {
+	// this is not a user-authenticated endpoint
+	svc.authz.SkipAuthorization(ctx)
+
+	host, err := svc.ds.LoadHostByOrbitNodeKey(ctx, orbitNodeKey)
+	if err != nil {
+		return nil, orbitError{
+			message:     "failed to find host by orbit node key: " + orbitNodeKey,
+			nodeInvalid: true,
+		}
+	}
```

Comment:
Given that this is an authenticated endpoint, we can do:
```go
host, ok := hostctx.FromContext(ctx)
if !ok {
	return nil, osqueryError{message: "internal error: missing host from request context"}
}
```
Because `authenticatedOrbitHost` does:
```go
ctx = hostctx.NewContext(ctx, host)
```